### PR TITLE
Bootstrap prefixes and ClassLoaderInstrumentation tweaks 

### PR DESCRIPTION
**Pull request by richardstartin**
* Makes it explicit that Datadog Classloaders aren't instrumented
* Enforces loading of bootstrap packages from bootstrap classloader when loaded by Datadog classloaders
* Exit early in `ClassLoaderInstrumentation` when the loaded class isn't a Datadog class so can't match by prefix

### CODE REVIEW

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/BootstrapLoadedPackages.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    "datadog.slf4j",
+    "datadog.trace.api",
+    "datadog.trace.bootstrap",
+    "datadog.trace.context",
+    "datadog.trace.instrumentation.api",
+    "datadog.trace.logging",
+    "datadog.trace.util",
+  };
+
+  public static Class<?> forceLoadWithBootstrapClassLoaderIfNecessary(String name) {
```
Comment (bantonsson) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/BootstrapLoadedPackages.java:29]: Nice that I'm not the only one with a love for descriptive method names

File: dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java [hunk 2]
```diff
@@ -42,7 +42,11 @@ public ClassloadingInstrumentation() {
   public ElementMatcher<TypeDescription> typeMatcher() {
     // just an optimization to exclude common class loaders that are known to delegate to the
     // bootstrap loader (or happen to _be_ the bootstrap loader)
-    return namedNoneOf("java.lang.ClassLoader", "com.ibm.oti.vm.BootstrapClassLoader")
+    return namedNoneOf(
+            "java.lang.ClassLoader",
+            "com.ibm.oti.vm.BootstrapClassLoader",
+            "datadog.trace.bootstrap.DatadogClassLoader",
+            "datadog.trace.bootstrap.DatadogClassLoader$DelegateClassLoader")
```
Comment (bantonsson) [inline dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java:49]: 👍

File: dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java [hunk 3]
```diff
@@ -65,10 +69,12 @@ public ClassloadingInstrumentation() {
...
   public static class LoadClassAdvice {
     @Advice.OnMethodEnter(skipOn = Advice.OnNonDefaultValue.class, suppress = Throwable.class)
-    public static Class<?> onEnter(@Advice.Argument(0) final String name) {
-      // we must access agent types used in the call-depth block like 'Constants' before entering it
-      // - otherwise we risk loading these agent types with a non-zero call-depth, which will fail
-      final String[] bootstrapPrefixes = Constants.BOOTSTRAP_PACKAGE_PREFIXES;
+    public static Class<?> onEnter(
+        @Advice.Argument(0) final String name, @Advice.Local("depth") int depth) {
+      // optimisation to avoid incrementing the call depth on every load
+      if (!name.startsWith("datadog.")) {
```
Comment (bantonsson) [inline dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java:75]: Shouldn't this logic live in the `BootstrapLoadedPackages` class?

**Review by bantonsson — APPROVED** [review]

File: dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java (hunk 3 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java:75]: It does, but that logic needs to be performed after the depth check, and I'm trying to avoid the depth check here, hence the comment

Comment (bantonsson) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java:75]: Yeah it's not a big thing, but I was more thinking of having a method there, so this would read something like:
```
if(!BootstrapLoadedPackages.shouldForceLoadWithBootstrapClassLoaderIfNecessary(name)) { 
  return null;
}
```

Comment (richardstartin) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java:75]: But the comment claims it's not safe to do that unless the depth check has been done, giving the excluded IBM `BootstrapClassloader` as an example of classloader which can create a circular dependency here.

Comment (richardstartin) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java:75]: OK I get it

Comment (richardstartin) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java]: PTAL

Comment (bantonsson) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java]: 👍

**Comment by richardstartin** [comment]
This currently doesn't support cases where bootstrap prefixed classes loaded through the datadog classloader can't be loaded by the by the bootstrap classloader, which will create a cycle.

File: dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java [hunk 4]
```diff
@@ -65,10 +70,12 @@ public ClassloadingInstrumentation() {
 
   public static class LoadClassAdvice {
     @Advice.OnMethodEnter(skipOn = Advice.OnNonDefaultValue.class, suppress = Throwable.class)
-    public static Class<?> onEnter(@Advice.Argument(0) final String name) {
-      // we must access agent types used in the call-depth block like 'Constants' before entering it
-      // - otherwise we risk loading these agent types with a non-zero call-depth, which will fail
```
Comment (mcculls) [inline dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java:70]: could we keep this comment in some form because it's useful information to have when making changes here

File: dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java [hunk 5]
```diff
@@ -65,10 +70,12 @@ public ClassloadingInstrumentation() {
...
-      final String[] bootstrapPrefixes = Constants.BOOTSTRAP_PACKAGE_PREFIXES;
+    public static Class<?> onEnter(
+        @Advice.Argument(0) final String name, @Advice.Local("depth") int depth) {
+      // optimisation to avoid incrementing the call depth on every load
+      if (!mayForceLoadWithBootstrapClassLoader(name)) {
```
Comment (mcculls) [inline dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java:76]: this also helps make sure `BootstrapLoadedPackages` has been loaded before we enter the call-depth block - normally it will be, but without a reference here there's a slim chance that an isolated classloader could call back on itself when trying to resolve the later `BootstrapLoadedPackages` reference and if the call-depth has already been bumped then that would fail...

File: dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java (hunk 4 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java:70]: I will reword it to make it more clear that really the problem is triggering class loads when the call depth is > 0

Comment (richardstartin) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java:70]: I would also be inclined to just load `Constants`/`BootstrapLoadedPackages` with the bootstrap classloader at startup to remove the risk entirely

Comment (mcculls) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java:70]: pre-loading won't necessarily help with isolated classloaders that use a different delegation model

File: dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java (hunk 5 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java:76]: I'm not sure I understand your point. How does the method call not ensure the class has been loaded? How could the method call return allowing the call depth to have been incremented without the class having been loaded?

Comment (richardstartin) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java]: I added a commit which calls the method through a singleton which requires a reference to have been loaded, but want to be convinced this is necessary https://github.com/DataDog/dd-trace-java/pull/2509/commits/dcbd17c46117f23544d392716128d8c56a735beb

Comment (mcculls) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java]: In the previous issue when we instrumented an isolated class-loader and injected the static field reference, if this wasn't done outside of the call-depth block then it triggered further class-loading. Moving that static field reference outside of the call-depth block was enough to avoid triggering this loop back. Method references may be treated differently but it's still a good idea IMHO to have a reference to `BootstrapLoadedPackages` outside of the call-depth block.

Comment (richardstartin) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java]: I understand why the class must be loaded before the call depth. What I don't understand is how the class could not have been loaded after the method call completes.

Comment (mcculls) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java]: A singleton isn't necessary here, I'm just pointing out that referencing the class outside of the call-depth block is enough - which is what my original comment was saying (ie. no changes required because of that class reference.)

There's a test in the codebase that attempts to recreate the original issue - I'll capture some logs as an example of what happened previously.

File: dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java (hunk 4 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java:70]: I've added a comment in any case

File: dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java (hunk 5 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java]: OK great, we agree then.

Comment (mcculls) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java]: I think we're talking cross-purposes here - I'm agreeing that the class must be loaded before the call depth increment, and that it will be with the code as it was in c06f1af (ie. you don't have to use a singleton)

I was just pointing out (more to others who might be reviewing along) that method call helps to load the class before the call-depth block, compared to not having that method call there.

**Review by mcculls — APPROVED** [review]
LGTM

File: dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java (hunk 5 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java:77]: The part I found ambiguous about this:

> this also helps make sure BootstrapLoadedPackages has been loaded before we enter the call-depth block - normally it will be, but without a reference here there's a slim chance that an isolated classloader could call back on itself when trying to resolve the later BootstrapLoadedPackages reference and if the call-depth has already been bumped then that would fail...

was 

> this also helps make sure BootstrapLoadedPackages has been loaded before we enter the call-depth block - *normally it will be, but without a reference here*
it's not really clearly what you mean by _it_. Does this mean the new code will normally ensure the class has been loaded, but not always, or does it mean the class will normally have been loaded? Then _without a reference here_... it's confusing whether you're saying the class needs to be loaded (which I think is the operative point) or that there are actually needs to be a reference load to avoid some kind of weird situation I haven't thought about, but that I cannot conceive of. Hence the back and forth.

I think the important point is that _the class must have been loaded whenever the depth is not zero_. We should also add a comment to `BootstrapLoadedPackages` to say that it must not load any non-bootstrap classes except during initialisation.

**Review by devinsba — APPROVED** [review]

File: dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java (hunk 5 above)
Comment (mcculls) [reply dd-java-agent/instrumentation/classloading/src/main/java/datadog/trace/instrumentation/classloading/ClassloadingInstrumentation.java:77]: > the class must have been loaded whenever the depth is not zero
Yes, that is the important point.

The case I was alluding to is a rare edge case that's easier to explain in a diagram - the code here won't fall into that edge case.

**Comment by richardstartin** [comment]
I'm not sure if the improvement here is worth the risk of the change.

**Review by tylerbenson — APPROVED** [review]
